sys/registry/registry.c

```diff
@@ -0,0 +1,198 @@
+    clist_rpush(&registry_handlers, &(handler->node));
+}
+
+static registry_handler_t *_handler_parse_and_lookup(char *name, int *name_argc,
+                                                     char *name_argv[])
+{
+    _parse_name(name, name_argc, name_argv);
+    return _handler_lookup(name_argv[0]);
+}
+
+int registry_set_value(char *name, char *val_str)
+{
+    int name_argc;
+    char *name_argv[REGISTRY_MAX_DIR_DEPTH];
+    registry_handler_t *hndlr;
+
+    hndlr = _handler_parse_and_lookup(name, &name_argc, name_argv);
+
+    if (!hndlr) {
+        return -EINVAL;
+    }
+
+    return hndlr->hndlr_set(name_argc - 1, &name_argv[1], val_str,
+                            hndlr->context);
+}
+
+char *registry_get_value(char *name, char *buf, int buf_len)
+{
+    int name_argc;
+    char *name_argv[REGISTRY_MAX_DIR_DEPTH];
+    registry_handler_t *hndlr;
+
+    hndlr = _handler_parse_and_lookup(name, &name_argc, name_argv);
+
+    if (!hndlr) {
+        return NULL;
+    }
+
+    if (!hndlr->hndlr_get) {
+        return NULL;
+    }
+
+    return hndlr->hndlr_get(name_argc - 1, &name_argv[1], buf, buf_len,
+                            hndlr->context);
+}
+
+static int _registry_call_commit(clist_node_t *current, void *res)
+{
+    assert(current != NULL);
+    int _res = *(int *)res;
+    registry_handler_t *hndlr = container_of(current, registry_handler_t, node);
+    if (hndlr->hndlr_commit) {
+        _res = hndlr->hndlr_commit(hndlr->context);
+        if (!*(int *)res) {
+            *(int *)res = _res;
+        }
+    }
+    return 0;
+}
+
+int registry_commit(char *name)
+{
+    int name_argc;
+    int rc = 0;
+
+    if (name) {
+        registry_handler_t *hndlr;
+        char *name_argv[REGISTRY_MAX_DIR_DEPTH];
+
+        hndlr = _handler_parse_and_lookup(name, &name_argc, name_argv);
+        if (!hndlr) {
+            return -EINVAL;
+        }
+        if (hndlr->hndlr_commit) {
+            return hndlr->hndlr_commit(hndlr->context);
+        }
+        else {
+            return 0;
+        }
+    }
+    else {
+        clist_foreach(&registry_handlers, _registry_call_commit,
+                      (void *)(&rc));
+        return rc;
+    }
+}
+
+int registry_export(int (*export_func)(const char *name, char *val), char *name)
+{
+    assert(export_func != NULL);
+    int name_argc;
+    char *name_argv[REGISTRY_MAX_DIR_DEPTH];
+    registry_handler_t *hndlr;
+
+    if (name) {
+        DEBUG("[registry export] exporting %s\n", name);
+        hndlr = _handler_parse_and_lookup(name, &name_argc, name_argv);
+        if (!hndlr) {
+            return -EINVAL;
+        }
+        if (hndlr->hndlr_export) {
+            return hndlr->hndlr_export(export_func, name_argc - 1,
+                                       &name_argv[1], hndlr->context);
+        }
+        else {
+            return 0;
+        }
+    }
+    else {
+        DEBUG("[registry export] exporting all\n");
+        clist_node_t *node = registry_handlers.next;
```

Comment:
I guess you do this because the `registry_handlers` is a node which doesn't represent any registry handler at all?
If that's the case, the `do` loop below will call `container_of` with the `registry_handlers` node at some point I think, is that right? Would that lead to an error?